velox/exec/Exchange.cpp

```diff
@@ -226,22 +210,27 @@ void ExchangeClient::noMoreRemoteTasks() {
 }
 
 void ExchangeClient::close() {
-  std::vector<std::shared_ptr<ExchangeSource>> sources;
-
+  if (closed_) {
+    return;
+  }
   {
-    std::lock_guard<std::mutex> l(queue_->mutex());
+    std::vector<std::shared_ptr<ExchangeSource>> sources;
+    {
+      std::lock_guard<std::mutex> l(queue_->mutex());
 
-    if (closed_) {
-      return;
+      closed_ = true;
+      sources = std::move(sources_);
     }
 
-    closed_ = true;
-    sources = std::move(sources_);
+    // Outside of mutex.
+    for (auto& source : sources) {
+      source->close();
+    }
   }
-
-  // Outside of mutex.
-  for (auto& source : sources) {
-    source->close();
+  {
+    std::shared_ptr<ExchangeQueue> queue;
+    queue = std::move(queue_);
+    VELOX_CHECK(queue.unique());
   }
```

Comment:
This is the actual fix.

Reply:
@tanjialiang 
This comment no longer actual as the code changed.